src/logging.c

```diff
@@ -0,0 +1,198 @@
+/*
+ * PgBouncer - Lightweight connection pooler for PostgreSQL.
+ *
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+/*
+ * Query logging
+ */
+
+#include "bouncer.h"
+
+#include <sys/file.h>
+#include <errno.h>
+
+#define LOG_BUFFER_SIZE 1024 * 1024 /* 1 MB */
+#define MAX_LOG_FILE_SIZE 1024 * 1024 * 25 /* 25 MB; if we get this far, the replayer isn't doing its job */
+
+/* Flush packets to log every 0.1 of a second */
+static struct timeval buffer_drain_period = {0, USEC / 10};
+static struct event buffer_drain_ev;
+
+/* The buffer */
+static char *buf;
+static size_t len;
+
+static void log_flush_buffer(void);
+void log_buffer_flush_cb(evutil_socket_t sock, short flags, void *arg);
+
+/*
+ * Initialize the packet logger.
+ */
+void log_init() {
+  /* Allocate the buffer */
+  buf = malloc(LOG_BUFFER_SIZE);
+  len = 0;
+
+  /* Flush the buffer every .1 of a second */
+  event_assign(&buffer_drain_ev, pgb_event_base, -1, EV_PERSIST, log_buffer_flush_cb, NULL);
+  event_add(&buffer_drain_ev, &buffer_drain_period);
```

Comment:
Similar things are not checked anywhere else in the code, I'm assuming because the author kinda guarantees it will be called only once / will not fail. No error handling here is possible really I think except `exit(1)`...